File: ee25btech11001/MATGEO/4.13.6/codes/code.c
#include <stdio.h>
#include <math.h>

// Function to compute inner product of two vectors
double inner_product(int n, double *u, double *v) {
    double sum = 0.0;
    for(int i=0; i<n; i++) {
        sum += u[i]*v[i];
    }
    return sum;
}

// Function to compute locus equation value at point P
// Equation: P^T (u u^T - D^2 I) P + (-(D^2+alpha)u + 2D^2 A)^T P + ((D^2+alpha)^2)/4 - D^2 A^T A
// Returns double value (should be 0 on the locus)
double locus_value(int n, double *A, double *B, double *P, double D) {
    // Compute u = A - B
    double u[10];  // assume dimension <= 10 for simplicity
    for(int i=0; i<n; i++) {
        u[i] = A[i] - B[i];
    }

    // Compute alpha = A^T A - B^T B
    double alpha = inner_product(n, A, A) - inner_product(n, B, B);

    // Compute P^T (u u^T - D^2 I) P
    double uuT_P[n];
    for(int i=0; i<n; i++) {
        uuT_P[i] = u[i]*inner_product(n, u, P);
    }
    double quad = inner_product(n, P, uuT_P);

    // Subtract D^2 * (P^T P)
    quad -= D*D * inner_product(n, P, P);

    // Compute linear term: (-(D^2+alpha)u + 2D^2 A)^T P
    double coeff[n];
    for(int i=0; i<n; i++) {
        coeff[i] = -(D*D+alpha)*u[i] + 2*D*D*A[i];
    }
    double lin = inner_product(n, coeff, P);

    // Constant term
    double constant = ((D*D+alpha)*(D*D+alpha))/4.0 - D*D*inner_product(n, A, A);

    return quad + lin + constant;
}

File: ee25btech11001/MATGEO/4.13.6/codes/code.py
import numpy as np
import matplotlib.pyplot as plt

def inner_product(u, v):
    return np.dot(u, v)

def locus_value(A, B, P, D):
    u = A - B
    alpha = inner_product(A, A) - inner_product(B, B)

    # Quadratic part
    quad = (np.dot(P, u))**2 - D*D*inner_product(P, P)

    # Linear part
    coeff = -(D*D+alpha)*u + 2*D*D*A
    lin = inner_product(coeff, P)

    # Constant part
    constant = ((D*D+alpha)**2)/4.0 - D*D*inner_product(A, A)

    return quad + lin + constant

# Example: A=(0,4), B=(0,-4), delta=6
A = np.array([0.0, 4.0])
B = np.array([0.0, -4.0])
D = 6.0

x = np.linspace(-10, 10, 400)
y = np.linspace(-10, 10, 400)
X, Y = np.meshgrid(x, y)
Z = np.zeros_like(X)

for i in range(X.shape[0]):
    for j in range(X.shape[1]):
        P = np.array([X[i,j], Y[i,j]])
        Z[i,j] = locus_value(A, B, P, D)

plt.contour(X, Y, Z, levels=[0], colors="red")
plt.axhline(0, color="k", linewidth=0.5)
plt.axvline(0, color="k", linewidth=0.5)
plt.gca().set_aspect("equal")
plt.title("Locus")
plt.show()
